bin/oneoff/backfill_data/migrate_workshop_survey_data.rb

```diff
@@ -0,0 +1,131 @@
+  "would_recommend_s" => "wouldRecommend",
+  "best_pd_ever_s" => "bestPdEver",
+  "part_of_community_s" => "partOfCommunity",
+  "things_you_liked_s" => "thingsYouLiked",
+  "things_you_would_change_s" => "thingsYouWouldChange",
+  "anything_else_s" => "anythingElse",
+  "gender_s" => "gender",
+  "race_ss" => "race",
+  "age_s" => "age",
+  "years_taught_i" => "yearsTaught",
+  "grades_taught_ss" => "gradesTaught",
+  "grades_planning_to_teach_ss" => "gradesPlanningToTeach",
+  "subjects_taught_ss" => "subjectsTaught",
+  "school_has_tech_b" => "schoolHasTech",
+  "venue_feedback_s" => "venueFeedback",
+  "how_heard_other_s" => "howHeardOther",
+  "will_not_teach_explanation_s" => "willNotTeachExplanation",
+  "willing_to_talk_b" => "willingToTalk",
+  "how_to_contact_s" => "howToContact",
+  "who_facilitated_ss" => "whoFacilitated"
+}
+
+NEW_FIELD_DEFAULTS = {
+  "whoFacilitated" => [],
+}
+
+# Custom class for creating Pd::WorkshopSurvey entries from older data that
+# might not necessarily pass our current validation
+class Pd::WorkshopSurveyAllowSomeNils < Pd::WorkshopSurvey
+  # These fields are new, so don't require them for older surveys
+  def required_fields
+    fields = super.dup
+    fields.delete(:school_has_tech)
+    fields.delete(:willing_to_talk)
+
+    fields
+  end
+
+  # because we can't expect the migration to enforce order and because we can
+  # trust the validation imposed by the old pegasus form to a certain extent,
+  # don't try to enforce demographics questions
+  def demographics_required_fields
+    []
+  end
+
+  # override the STI name because we want these to be saved as regular old
+  # Pd::WorkshopSurvey instances, just having gone through our custom validation
+  def self.sti_name
+    Pd::WorkshopSurvey.sti_name
+  end
+end
+
+def main
+  PEGASUS_DB[:forms].where(kind: 'PdWorkshopSurvey').each do |survey|
+    old_data = JSON.parse survey[:data]
+
+    # If this survey is one of the few for which the user did not consent,
+    # simply discard it
+    next if old_data["consent_b"] == "0"
+
+    pd_enrollment_id = old_data["enrollment_id_i"].to_i
+
+    # we enforce uniqueness of the enrollment id at all levels, so fail fast
+    # here if we've already created a survey from this data. This will allow us
+    # to rerun this script
+    next if Pd::WorkshopSurvey.exists?(pd_enrollment_id: pd_enrollment_id)
+
+    survey = Pd::WorkshopSurveyAllowSomeNils.create(pd_enrollment_id: pd_enrollment_id)
+
+    new_data = {}
+    old_data.each do |key, value|
+      new_data[FIELD_MAPPING[key]] = value if FIELD_MAPPING.key?(key) && !value.nil?
+    end
+
+    # boolean fields need to be converted to text
+    booleans = %w(willTeach schoolHasTech willingToTalk)
+    booleans.each do |boolean|
+      next unless new_data.key? boolean
+      new_data[boolean] = new_data[boolean] == "1" ? "Yes" : "No"
+    end
+
+    # integer fields need to be converted to text
+    integers = %w(yearsTaught)
```

Comment:
This confuses me. Why are we creating an array of one element, then iterating over the array, rather than simply doing `new_data['yearsTaught'] = (new_data.try(:[], 'yearsTaught') || 0).to_s`?